crates/typst-ide/src/matchers.rs

```diff
@@ -0,0 +1,178 @@
+            if let Some(v) = parent.cast::<ast::ForLoop>() {
+                if node.prev_sibling_kind() != Some(SyntaxKind::In) {
+                    let pattern = v.pattern();
+                    for ident in pattern.bindings() {
+                        if let Some(res) = recv(NamedItem::Var(ident)) {
+                            return Some(res);
+                        }
+                    }
+                }
+            }
+
+            ancestor = Some(parent.clone());
+            continue;
+        }
+
+        break;
+    }
+
+    None
+}
+
+pub enum NamedItem<'a> {
+    Var(ast::Ident<'a>),
+    Fn(ast::Ident<'a>),
+    Module(&'a Module),
+    Import(&'a EcoString, Span, Option<&'a Value>),
+}
+
+impl<'a> NamedItem<'a> {
+    pub(crate) fn name(&self) -> &'a EcoString {
+        match self {
+            NamedItem::Var(ident) => ident.get(),
+            NamedItem::Fn(ident) => ident.get(),
+            NamedItem::Module(value) => value.name(),
+            NamedItem::Import(name, _, _) => name,
+        }
+    }
+
+    pub(crate) fn value(&self) -> Option<Value> {
+        match self {
+            NamedItem::Var(..) | NamedItem::Fn(..) => None,
+            NamedItem::Module(value) => Some(Value::Module((*value).clone())),
+            NamedItem::Import(_, _, value) => value.cloned(),
+        }
+    }
+}
+
+/// Find an expression that can be dereferenced on the given node.
+pub fn deref_target(node: LinkedNode) -> Option<DerefTarget<'_>> {
+    // Move to the first ancestor that is an expression.
+    let mut ancestor = node;
+    while !ancestor.is::<ast::Expr>() {
+        ancestor = ancestor.parent()?.clone();
+    }
+
+    // Identify convenient expression kinds.
+    let expr_node = ancestor;
+    let expr = expr_node.cast::<ast::Expr>()?;
+    Some(match expr {
+        ast::Expr::Label(..) => DerefTarget::Label(expr_node),
+        ast::Expr::Ref(..) => DerefTarget::Ref(expr_node),
+        ast::Expr::FuncCall(call) => {
+            DerefTarget::Callee(expr_node.find(call.callee().span())?)
+        }
+        ast::Expr::Set(set) => DerefTarget::Callee(expr_node.find(set.target().span())?),
+        ast::Expr::Ident(..) | ast::Expr::MathIdent(..) | ast::Expr::FieldAccess(..) => {
+            DerefTarget::VarAccess(expr_node)
+        }
+        ast::Expr::Str(..) => {
+            let parent = expr_node.parent()?;
+            if parent.kind() == SyntaxKind::ModuleImport {
+                DerefTarget::ImportPath(expr_node)
+            } else if parent.kind() == SyntaxKind::ModuleInclude {
+                DerefTarget::IncludePath(expr_node)
+            } else {
+                DerefTarget::Normal(expr_node.kind(), expr_node)
+            }
+        }
+        _ if expr.hash()
+            || matches!(expr_node.kind(), SyntaxKind::MathIdent | SyntaxKind::Error) =>
+        {
+            DerefTarget::Normal(expr_node.kind(), expr_node)
+        }
+        _ => return None,
```

Comment:
What is the exact definition of `Normal` vs `None`. Why `hash()`?

Reply:
It isn't well defined. The condition to check is borrowed on some existing code:

```rs
let mut ancestor = leaf;
while !ancestor.is::<ast::Expr>() {
    ancestor = ancestor.parent()?;
}

let expr = ancestor.cast::<ast::Expr>()?;
if !expr.hash() && !matches!(expr, ast::Expr::MathIdent(_)) {
    return None;
}
```

Perhaps some markup expressions are excluded, but I'm not quite sure an accurate list of them.

https://github.com/Myriad-Dreamin/typst/blob/3e1156787f1f05c514a80f3a7fa237f3dbf5c8d5/crates/typst-ide/src/tooltip.rs#L51-L59

Reply:
ah okay, yeah, there it excludes markup expression because that would get annoying real quick on hover